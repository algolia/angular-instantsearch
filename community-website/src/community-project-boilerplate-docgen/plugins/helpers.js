const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
