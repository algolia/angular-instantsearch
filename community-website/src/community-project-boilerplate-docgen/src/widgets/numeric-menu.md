---
title: Widgets - Numeric Menu
layout: widget.pug
---

# Numeric Menu

## Description

The Numeric Menu is a component that displays a list of numeric filters in a list. Those numeric filters are pre-configured with creating the widget.

## Options

| Attribute            | Type                                             | Description
| -                    | -                                                | -
| `attribute         ` | `string`                                         | Name of the numeric attribute to use
| `items`              | `{name: string, start?: number, end?: number}[]` | The list of indexes to search in
| `autoHideContainer?` | `boolean`  | Hides the numeric menu if there's no item to display

## Code example

```html
<ais-numeric-menu
  attribute="price"
  operator="or"
  [items]="[
    { name: 'All' },
    { end: 4, name: 'less than 4' },
    { start: 4, end: 4, name: '4' },
    { start: 5, end: 10, name: 'between 5 and 10' },
    { start: 10, name: 'more than 10' }
  ]"
>
</ais-numeric-menu>
```
