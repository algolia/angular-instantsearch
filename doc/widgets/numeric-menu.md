## `<ng-ais-numeric-menu />`

The NumericMenu is a component that displays a list of numeric filters in a list. Those numeric filters are pre-configured with creating the widget.

| Attributes           | Type                                             | Description
| -                    | -                                                | -
| `attribute         ` | `string`                                         | Name of the numeric attribute to use
| `options`            | `{name: string, start?: number, end?: number}[]` | The list of indexes to search in
| `autoHideContainer?` | `boolean`  | Hides the numeric menu if there's no item to display
