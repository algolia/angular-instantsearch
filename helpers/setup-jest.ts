import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
