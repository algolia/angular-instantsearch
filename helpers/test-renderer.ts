import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
