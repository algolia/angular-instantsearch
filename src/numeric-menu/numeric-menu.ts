import { Component, Input, Inject, forwardRef } from "@angular/core";

import { connectNumericRefinementList } from "instantsearch.js/es/connectors";
import { noop } from "lodash-es";

import { BaseWidget } from "../base-widget";
import { NgAisInstantSearch } from "../instantsearch/instantsearch";

export type NumericRefinementListState = {
  createURL: Function;
  items: {}[];
  refine: Function;
};

@Component({
  selector: "ng-ais-numeric-menu",
  template: `
    <div
      [class]="cx()"
      *ngIf="!isHidden"
    >
      <ul [class]="cx('list')">
        <li
          [class]="getItemClass(item)"
          *ngFor="let item of state.items"
          (click)="refine($event, item)"
        >
          <label [class]="cx('label')">
            <input
              [class]="cx('radio')"
              type="radio"
              name="NumericMenu"
              [checked]="item.isRefined"
            />
            <span [class]="cx('labelText')">{{item.label}}</span>
          </label>
        </li>
      </ul>
    </div>
  `
})
export class NgAisNumericMenu extends BaseWidget {
  @Input() public attribute: string;
  @Input()
  public options: {
    name: string;
    start?: number;
    end?: number;
  }[];

  public state: NumericRefinementListState = {
    createURL: noop,
    items: [],
    refine: noop
  };

  get isHidden() {
    return this.state.items.length === 0 && this.autoHideContainer;
  }

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super("NumericMenu");
  }

  public ngOnInit() {
    this.createWidget(connectNumericRefinementList, {
      attributeName: this.attribute,
      options: this.options
    });
    super.ngOnInit();
  }

  public refine(event: MouseEvent, item: { value: string }) {
    event.preventDefault();
    event.stopPropagation();
    this.state.refine(item.value);
  }
}
