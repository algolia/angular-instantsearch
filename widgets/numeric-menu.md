## `<ng-ais-numeric-menu />`

The NumericMenu is a component that displays a list of numeric filters in a list. Those numeric filters are pre-configured with creating the widget.

| Attributes      | Type                                                 | Description
| -               | -                                                | -
| `attributeName` | `string`                                         | Name of the numeric attribute to use
| `options`       | `{name: string, start?: number, end?: number}[]` | The list of indexes to search in
| `header?`       | `string`                                         | Displays text before widget
| `footer?`       | `string`                                         | Displays text after widget
