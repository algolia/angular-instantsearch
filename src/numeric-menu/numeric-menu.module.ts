import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisNumericMenu } from './numeric-menu';

@NgModule({
  declarations: [NgAisNumericMenu],
  entryComponents: [NgAisNumericMenu],
  exports: [NgAisNumericMenu],
  imports: [CommonModule],
})
export class NgAisNumericMenuModule {}
