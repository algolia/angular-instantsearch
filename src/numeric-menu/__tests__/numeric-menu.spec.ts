import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisNumericMenu } from "../numeric-menu";

const defaultState = {
  createURL: jest.fn(),
  items: [
    { label: "10", value: "10", isRefined: true },
    { label: "20", value: "20" }
  ],
  refine: jest.fn()
};

const render = createRenderer({
  defaultState,
  template: "<ng-ais-numeric-menu></ng-ais-numeric-menu>",
  TestedWidget: NgAisNumericMenu
});

describe("NumericRefinementList", () => {
  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should refine() when item click", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const [, item] = fixture.debugElement.nativeElement.querySelectorAll("li");
    item.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultState.items[1].value);
  });

  it("should check the input of item.isRefined", () => {
    const fixture = render({});
    const [
      firstInput,
      secondInput
    ] = fixture.debugElement.nativeElement.querySelectorAll("input");

    expect(firstInput.checked).toBeTruthy();
    expect(secondInput.checked).toBeFalsy();
  });

  it("should be hidden with autoHideContainer", () => {
    const fixture = render({ items: [] });
    fixture.componentInstance.testedWidget.autoHideContainer = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
