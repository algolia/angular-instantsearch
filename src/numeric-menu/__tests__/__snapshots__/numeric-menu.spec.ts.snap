// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericRefinementList renders markup with state 1`] = `
<ng-ais-numeric-menu
  cx={[Function Function]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-NumericMenu"
  >
    
        
    
    <div
      class="ais-NumericMenu-body ais-body"
    >
      
          
      <ul
        class="ais-NumericMenu-list"
      >
        
            
        
        <li
          class="ais-NumericMenu-item ais-NumericMenu-item--selected"
        >
          
              
          <label
            class="ais-NumericMenu-label"
          >
            
                
            <input
              class="ais-NumericMenu-radio"
              name="NumericRefinementList"
              type="radio"
            />
            
                10
              
          </label>
          
            
        </li>
        <li
          class="ais-NumericMenu-item"
        >
          
              
          <label
            class="ais-NumericMenu-label"
          >
            
                
            <input
              class="ais-NumericMenu-radio"
              name="NumericRefinementList"
              type="radio"
            />
            
                20
              
          </label>
          
            
        </li>
        
          
      </ul>
      
        
    </div>
    
      
  </div>
    
</ng-ais-numeric-menu>
`;

exports[`NumericRefinementList renders markup without state 1`] = `
<ng-ais-numeric-menu
  cx={[Function Function]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-NumericMenu"
  >
    
        
    
    <div
      class="ais-NumericMenu-body ais-body"
    >
      
          
      <ul
        class="ais-NumericMenu-list"
      >
        
            
        
        
          
      </ul>
      
        
    </div>
    
      
  </div>
    
</ng-ais-numeric-menu>
`;

exports[`NumericRefinementList should be hidden with autoHideContainer 1`] = `
<ng-ais-numeric-menu
  autoHideContainer={[Function Boolean]}
  cx={[Function Function]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-NumericMenu"
  >
    
        
    
    
      
  </div>
    
</ng-ais-numeric-menu>
`;
